File: articles/calculus-for-systems-modeling/articles/gradient-divergence-and-curl/fortran/field_operator_audit.f90
program field_operator_audit
  implicit none
  print '(A)', 'scenario grid_step point_count mean_gradient_magnitude maximum_gradient_magnitude mean_divergence mean_curl maximum_abs_curl field_description warning'
  call audit(1.0d0, 'coarse_grid')
  call audit(0.5d0, 'medium_grid')
  call audit(0.25d0, 'fine_grid')
contains
  subroutine gradient(x,y,gx,gy)
    real(8), intent(in) :: x,y
    real(8), intent(out) :: gx,gy
    gx = 2.0d0*x
    gy = 2.0d0*y
  end subroutine

  subroutine audit(step, scenario)
    real(8), intent(in) :: step
    character(len=*), intent(in) :: scenario
    integer :: i,j,n,count
    real(8) :: x,y,gx,gy,gmag,grad_sum,max_grad,div_sum,curl_sum,max_abs_curl,curl
    character(len=32) :: warning
    n = int(2.0d0 / step) + 1
    count = 0
    grad_sum = 0.0d0
    max_grad = 0.0d0
    div_sum = 0.0d0
    curl_sum = 0.0d0
    max_abs_curl = 0.0d0
    do i = 0, n-1
      x = -1.0d0 + i*step
      do j = 0, n-1
        y = -1.0d0 + j*step
        call gradient(x,y,gx,gy)
        gmag = sqrt(gx*gx + gy*gy)
        curl = 2.0d0
        count = count + 1
        grad_sum = grad_sum + gmag
        max_grad = max(max_grad, gmag)
        div_sum = div_sum + 0.0d0
        curl_sum = curl_sum + curl
        max_abs_curl = max(max_abs_curl, abs(curl))
      end do
    end do
    if (step > 0.5d0) then
      warning = 'coarse_grid'
    else
      warning = 'synthetic_field'
    end if
    print '(A,1X,F8.3,1X,I8,5F14.6,1X,A,1X,A)', trim(scenario), step, count, grad_sum/count, max_grad, div_sum/count, curl_sum/count, max_abs_curl, 'field', trim(warning)
  end subroutine
end program

File: articles/calculus-for-systems-modeling/articles/gradient-divergence-and-curl/c/field_operator_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double scalar_field(double x, double y){ return x*x + y*y; }
void vector_field(double x, double y, double* p, double* q){ *p = -y; *q = x; }
void gradient(double x, double y, double* gx, double* gy){ *gx = 2.0*x; *gy = 2.0*y; }
double divergence(double x, double y){ (void)x; (void)y; return 0.0; }
double curl_2d(double x, double y){ (void)x; (void)y; return 2.0; }

void audit(double step, const char* scenario){
  int n = (int)(2.0 / step) + 1;
  int count = 0;
  double grad_sum = 0.0, max_grad = 0.0, div_sum = 0.0, curl_sum = 0.0, max_abs_curl = 0.0;
  for(int i=0; i<n; i++){
    double x = -1.0 + i*step;
    for(int j=0; j<n; j++){
      double y = -1.0 + j*step;
      double gx, gy;
      gradient(x,y,&gx,&gy);
      double gmag = sqrt(gx*gx + gy*gy);
      double div = divergence(x,y);
      double curl = curl_2d(x,y);
      count++;
      grad_sum += gmag;
      if(gmag > max_grad) max_grad = gmag;
      div_sum += div;
      curl_sum += curl;
      if(fabs(curl) > max_abs_curl) max_abs_curl = fabs(curl);
    }
  }
  const char* warning = step > 0.5 ? "Grid step is coarse; local derivative structure may be undersampled." : "Synthetic field-operator audit; document field definitions units grid and boundary rules.";
  printf("%s,%.12f,%d,%.12f,%.12f,%.12f,%.12f,%.12f,%s,%s\n", scenario, step, count, grad_sum/count, max_grad, div_sum/count, curl_sum/count, max_abs_curl, "scalar f=x^2+y^2; vector F=<-y,x>", warning);
}

int main(void){
  printf("scenario,grid_step,point_count,mean_gradient_magnitude,maximum_gradient_magnitude,mean_divergence,mean_curl,maximum_abs_curl,field_description,warning\n");
  audit(1.0, "coarse_grid");
  audit(0.5, "medium_grid");
  audit(0.25, "fine_grid");
  return EXIT_SUCCESS;
}
